### ISSUE TITLE
Refactor sensor state in Vernier Force and Acceleration extension

### ISSUE THREAD

**Issue by ericrosenbaum** [LLK/scratch-vm#1953]
In our other hardware extensions,  the `_peripheral._sensors` object is initialized with default sensor values at extension load. Data from the hardware directly updates these values. The block opcodes can then simply use `_peripheral`'s getters, which are always safe (i.e. whether or not the hardware is connected). 

In the vernier extension, we currently have a more complex setup: `_peripheral` has a `_device` created by the Vernier godirect module, which has internal state tracking the sensor values. The `_device` object only exists after the hardware has been connected. This situation necessitated adding both a boolean _sensorsEnabled and a function _canReadSensors(), which is used in every sensor value getter and requires the “default” state of sensor values to be distributed throughout the code. This setup seems complex and error prone.

I think we should change it to be more like our other extensions. The `_peripheral` should have a `_sensors` object maintaining the sensor values, which listens for sensor update events from the hardware. The `_sensors` object just needs to be reset to defaults on load and on disconnect.

**Comment by evhan55** [comment]
This sounds worth doing if it maintains or fixes current behavior.
Are there reproducible errors that you are encountering?

I have thought about making something like `ScratchLinkDeviceAdapter` for all extensions at some point, but was waiting for the various connect/disconnect and read/write patterns among all the existing hardware extensions to reach a stable and robust state before adding another abstraction.

My vote would be:
1. Make the `GDX-FOR` extension more closely match other ones for now, especially if it will help with random bugs or edge cases we are encountering
2. Add TaskQueue (rate limiting) to all four hardware extensions: micro:bit/EV3/WeDo2/GDX_FOR
3. Verify that connect/disconnect/write/read/rate-limiting are feeling robust for all four extensions
4. Try factoring out something similar to `ScratchLinkDeviceAdapter` that is used by all hardware extensions, possibly merging in with and combining `io/BT` and `io/BLE`
5. Write tests for all 4 extensions

It would be nice to try to do some or all of the above steps before adding in too many more hardware extensions, imo.

**Comment by ericrosenbaum** [comment]
@evhan55 that sounds like a good plan!

I believe that the status of `ScratchLinkDeviceAdapter` is not tightly coupled to the specific refactor I described above (the changes I described would not touch that adapter). At some point you had suggested (I think) renaming it to something like `VernierGoDirectAdapter` (because it is adapting the Vernier Go Direct module so it can work with Scratch Link, and not the other way around). I think that might make the situation clearer. The Vernier extension is different from the others because it uses the Vernier Module, and so it requires that adapter, but maybe regardless we should factor out that connection logic from the other extensions like you said.
